pandas/_libs/interval.pyi

```diff
@@ -0,0 +1,141 @@
+from __future__ import annotations
+
+import sys
+from typing import (
+    Any,
+    Generic,
+    Protocol,
+    TypeVar,
+    Union,
+    overload,
+)
+
+import numpy as np
+
+from pandas._typing import (
+    Timedelta,
+    Timestamp,
+)
+
+if sys.version_info >= (3, 8):
+    from typing import Literal
+else:
+    from typing_extensions import Literal
+
+VALID_CLOSED: frozenset
+
+OrderableScalarT = TypeVar("OrderableScalarT", int, float)
+OrderableTimesT = TypeVar("OrderableTimesT", Timestamp, Timedelta)
+OrderableT = TypeVar("OrderableT", int, float, Timestamp, Timedelta)
+
+class IntervalMixinProtocol(Protocol): ...
+
+class _LengthDescriptor:
+    @overload
+    def __get__(self, instance: Interval[float], owner: Any) -> float: ...
+    @overload
+    def __get__(self, instance: Interval[int], owner: Any) -> int: ...
+    @overload
+    def __get__(self, instance: Interval[OrderableTimesT], owner: Any) -> Timedelta: ...
+
+class _MidDescriptor:
+    @overload
+    def __get__(self, instance: Interval[OrderableScalarT], owner: Any) -> float: ...
+    @overload
+    def __get__(self, instance: Interval[Timedelta], owner: Any) -> Timedelta: ...
+    @overload
+    def __get__(self, instance: Interval[Timestamp], owner: Any) -> Timestamp: ...
```

Comment:
You can probably collapse the last two overloads into one (which also takes care of sub-classes):

```py
@overload
def __get__(self, instance: Interval[OrderableTimesT], owner: Any) -> OrderableTimesT: ...
```

You will also need an overload for IntervalTree (for `mid/length`).

Reply:
done in next commit